### PR TITLE
Fastsim newrechits rebase sep4

### PR DESCRIPTION
**Pull request by lveldere**
rebase of #11078
- fastsim tracker rechit classes are replaced with
  - a fastsim base class FastTrackerRecHit
  - a fastsim class for single rechits FastSingleTrackerRecHit
    (accomodates the fastsim equivalents of SiPixelRecHit, SiStripRecHit2D and SiStripRecHit1D)
  - a fastsim equivalent of SiStripMatchedRecHit2D, FastMatchedTrackerRecHit
  - a fastsim equivalent of ProjectedSiStripRecHit2D, FastProjectedTrackerRecHit
- trackerHitRTTI::RTTI is extended to accommodate the new rechit containers
- a new implementation of FastTrackerRecHit::sharesInput
  ( will receive further study once merged)
- rechit production is further modularised
  - actual rechit production by the old SiTrackerGaussianSmearingRecHitConverter
  - hit matching by new FastTrackerRecHitMatcher
  - predefined combinations of hits by new FastTrackerRecHitCombiner
- SiTrackerGaussianSmearingRecHitConverter produces now
  OwnVector<FastTrackerRecHit>
  vector<\Ref<FastTrackerRecHitCollection> > with one entry per SimHit to find for each SimHit the corresponding RecHit
- predefined combinations of hits are now in a new, simple format:
  vector<vector<Ref<FastTrackerRecHitCollection> > >
- model for hit splitting has been improved and moved to a separate class FastTrackerRecHitSplitter
- model for masking of hits in iterative tracking improved
  (see HitMaskHelper and FastTrackerRecHitMaskProducer)
- GlobalPixelTracking is not fully migrated yet (it produces zero result):
  it was not properly configured to begin with,
  and we need to study the fullsim equivalent before fixing it
- various code files were updated for the new rechit classes and the new rechit combination format
- y position of strip hits is set to 0 (before it was the simhit y-position)
- unit tests: a few fastsim events are generated on the fly, such to circumvent the backward incompatibility introduced by this pr. Unit tests to be restored when new relvals are available.

### CODE REVIEW

**Comment by lveldere** [comment]
please test

**Comment by lveldere** [comment]
+1 

fullsim comparisons look okay
fastsim changes are expected

@civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @monttj, @slava77, @vadler
It would be extremely helpful if you could treat this with some priority.
I only have a few working days left before holidays...

**Comment by civanch** [comment]
+1

File: FastSimulation/Tracking/plugins/TrackCandidateProducer.cc [hunk 1]
```diff
@@ -113,44 +113,64 @@ TrackCandidateProducer::produce(edm::Event& e, const edm::EventSetup& es) {
...
+      continue;
     }
 
     // Get the combination of hits that produced the seed
-    int32_t hitCombinationId =  ((const SiTrackerGSMatchedRecHit2D*) (&*(seed.recHits().first)))->hitCombinationId();
-    const FastTMRecHitCombination & recHitCombination = recHitCombinations->at(hitCombinationId);
+    int32_t icomb = fastTrackingHelper::getRecHitCombinationIndex(seed);
+    if(icomb < 0 || unsigned(icomb) >= recHitCombinations->size()){
+	edm::LogError("TrackCandidateProducer") << " found seed with recHitCombination out or range: " << icomb << std::endl;
+	exit(1);
```
Comment (slava77) [inline FastSimulation/Tracking/plugins/TrackCandidateProducer.cc]: this should be removed and replaced with an exception

File: FastSimulation/TrackingRecHitProducer/plugins/FastTrackerRecHitMatcher.cc [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+    
+    StripPosition project(const GeomDetUnit *det,
+			  const GluedGeomDet* glueddet,
+			  const StripPosition& strip,
+			  const LocalVector& trackdirection)const;
+    
+    inline const FastSingleTrackerRecHit * _cast2Single(const FastTrackerRecHit * recHit) const{
+	if(!recHit->isSingle()){
+	    edm::LogError("FastTrackerRecHitMatcher") << "all rechits in simHit2RecHitMap must be instances of FastSingleTrackerRecHit. recHit's rtti: " << recHit->rtti() << std::endl;
+	    exit(1);
```
Comment (slava77) [inline FastSimulation/TrackingRecHitProducer/plugins/FastTrackerRecHitMatcher.cc]: no system calls/exits please: use cms::Exception

File: FastSimulation/Tracking/interface/FastTrackingHelper.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	    static_cast<FastTrackerRecHit &>(recHit).setRecHitCombinationIndex(icomb);
+	}
+    }
+
+    // get recHitCombination for TrackCandidate and TrajectorySeed
+    template<class T> int32_t getRecHitCombinationIndex(const T & object){
+	// seed must have at least one hit
+	if(object.recHits().first == object.recHits().second){
+	    edm::LogError("fastTrackingHelpers::getRecHitCombinationIndex") << "  given object has 0 hits" << std::endl;
+	    exit(1);
```
Comment (slava77) [inline FastSimulation/Tracking/interface/FastTrackingHelper.h]: one more

**Comment by slava77** [comment]
+1

- code changes are mostly OK
    - *exit(1) system calls should still be removed * (this can be done in a separate PR)
    - fullsim/data reco is not affected
- jenkins test pass and comparisons with baseline show differences only in the fastsim workflows
- I scrolled through a large fraction of comparison plots in 135.4 (ZEE fastsim) and the distributions changed mainly in the tracking algorithm side and less so in the general track or other variables properties
wf135p4_gentrk_algo
wf135p4_gentrk_hits_v_eta
- logfiles still suggest there is a problem in the TrajectorySeedProducer:pixelTripletSeeds:
this message changed from something that appears frequently and is filling up the log file
``` 
Either region producer, or the seed creator cfg is not available.
```
to something that shows up less frequently
```
 RegionFactory is not initialised properly, please check your configuration. Producing empty seed collection
```

**Comment by lveldere** [comment]
@slava77 
Thanks for the comments!
I will get rid of the exits as soon as this pr gets merged.
About the log files: this is a known issue in the HLT paths.
I have a branch with a fix, and I will make a pr when this one gets merged.

**Comment by Dr15Jones** [comment]
@davidlange6 The removal of the class ```SiTrackerGSMatchedRecHit2D``` is causing the IB RelVals to fail with the message
```
----- Begin Fatal Exception 11-Sep-2015 10:25:21 CEST-----------------------
An exception of category 'FileReadError' occurred while
   [0] Processing run: 1 lumi: 1 event: 1
   [1] Running path 'datamixing_step'
   [2] Calling event method for module DataMixingModule/'mixData'
   [3] Reading branch TrackingRecHitsOwned_mix_generalTracks_DIGI2RAW.
   Additional Info:
      [a] Fatal Root Error: @SUB=TBufferFile::ReadObject
trying to read an emulated class (SiTrackerGSMatchedRecHit2D) to store in a compiled pointer (TrackingRecHit)

----- End Fatal Exception -------------------------------------------------
```

see
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc493/CMSSW_7_6_X_2015-09-10-2300/pyRelValMatrixLogs/run/250400.0_ZEE_13+FS_ZEE_13_PRMXUP15_PU25+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+FS_ZEE_13_PRMXUP15_PU25+HARVESTUP15FS+MINIAODMCUP15FS.log

**Pull request merged**